# Week 5 Assignment IS 362

This notebook answers the following three questions:

1. What is the northernmost airport in the United States?
2. What is the easternmost airport in the United States?
3. On February 12th, 2013, which New York area airport had the windiest weather?

In [1]:
import pandas as pd

airports_url = "https://raw.githubusercontent.com/tidyverse/nycflights13/master/data-raw/airports.csv"
weather_url = "https://raw.githubusercontent.com/tidyverse/nycflights13/master/data-raw/weather.csv"

airports = pd.read_csv(airports_url)
weather = pd.read_csv(weather_url)

# show the first few rows of each to confirm load
print(airports.head())
print(weather.head())

   faa                           name        lat        lon   alt  tz dst  \
0  04G              Lansdowne Airport  41.130472 -80.619583  1044  -5   A   
1  06A  Moton Field Municipal Airport  32.460572 -85.680028   264  -6   A   
2  06C            Schaumburg Regional  41.989341 -88.101243   801  -6   A   
3  06N                Randall Airport  41.431912 -74.391561   523  -5   A   
4  09J          Jekyll Island Airport  31.074472 -81.427778    11  -5   A   

              tzone  
0  America/New_York  
1   America/Chicago  
2   America/Chicago  
3  America/New_York  
4  America/New_York  
  origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
0    EWR  2013      1    1     1  39.02  26.06  59.37     270.0    10.35702   
1    EWR  2013      1    1     2  39.02  26.96  61.63     250.0     8.05546   
2    EWR  2013      1    1     3  39.02  28.04  64.43     240.0    11.50780   
3    EWR  2013      1    1     4  39.92  28.04  62.21     250.0    12.65858   
4    EWR  

# Question 1: What is the northernmost airport in the United States?

In [2]:
# find the five northernmost airports by latitude
# sort by latitude in descending order to get highest latitudes first
top5_north = airports.sort_values(by="lat", ascending=False).head(5)

# display the top 5 northernmost airports with lat lon
print(top5_north[['name', 'lat', 'lon']])

                                           name        lat         lon
417                     Dillant Hopkins Airport  72.270833   42.898333
230                  Wiley Post Will Rogers Mem  71.285446 -156.766003
110                          Wainwright Airport  70.638056 -159.994722
708                               Wainwright As  70.613378 -159.860350
152  Atqasuk Edward Burnell Sr Memorial Airport  70.467300 -157.436000


Sorting the data by descending latitude displays Dillant Hopkins Airport with the highest latitude. However, a quick Google search shows that Dillant Hopkins Airport is in New Hampshire. Its coordinates are reversed.

In [3]:
airports.loc[417, 'lat'] = 42.898333
airports.loc[417, 'lon'] = -72.270833
print (airports.loc[417][['name', 'lat', 'lon']])

name    Dillant Hopkins Airport
lat                   42.898333
lon                  -72.270833
Name: 417, dtype: object


The latitude and longitude vales for Dillant Hopkins Airport has been corrected using 'pandas.loc'.

In [4]:
top5_north = airports.sort_values(by="lat", ascending=False).head(5)
print(top5_north[['name', 'lat', 'lon']])

                                            name        lat         lon
230                   Wiley Post Will Rogers Mem  71.285446 -156.766003
110                           Wainwright Airport  70.638056 -159.994722
708                                Wainwright As  70.613378 -159.860350
152   Atqasuk Edward Burnell Sr Memorial Airport  70.467300 -157.436000
1363                        Ugnu-Kuparuk Airport  70.330833 -149.597500


With the dataset corrected, the analysis now shows the correct 5 northernmost airports.

The answer to question 1 is **Wiley Post Will Rogers Memorial Airport** in Alaska, with a latitude of 71.29.

# Question 2: What is the easternmost airport in the United States?

In [5]:
#filter out non-negative numbers
airports_clean = airports[airports['lon']<0]
# find the five easternmost airports by longitude
top5_east = airports_clean.sort_values(by="lon", ascending=False).head(5)

# display the top 5 easternmost airports with lat lon
top5_east[["faa", "name", "lat", "lon"]]


faa                                 name        lat        lon
444   EPM           Eastport Municipal Airport  44.910111 -67.012694
624   HUL                         Houlton Intl  46.123083 -67.792056
259   CAR                         Caribou Muni  46.871500 -68.017917
1101  PQI  Northern Maine Rgnl At Presque Isle  46.688958 -68.044797
1398  WFK  Northern Aroostook Regional Airport  47.285556 -68.312778

After filtering out incorrect longitude values (such as airports missing negative signs), the five easternmost airports in the United States were identified by sorting for the largest negative longitude values.  

The answer to question 2 is **Eastport Municipal Airport (EPM)** in Maine, with a longitude of -67.01. 

# Question 3: On February 12th, 2013, which New York area airport had the windiest weather?

In [6]:
# Checking the columns in the weather.csv
print (weather.columns)


Index(['origin', 'year', 'month', 'day', 'hour', 'temp', 'dewp', 'humid',
       'wind_dir', 'wind_speed', 'wind_gust', 'precip', 'pressure', 'visib',
       'time_hour'],
      dtype='object')


In [7]:
# looking only at the weather on Feb 12, 2013
w = weather[(weather.year == 2013) & (weather.month == 2) & (weather.day == 12)].copy()
print (w.head())

# checking which airports have data on that date
print ("Airports: ", w["origin"].unique())

     origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
1006    EWR  2013      2   12     0  39.92  39.02  96.55     240.0   
1007    EWR  2013      2   12     1  39.92  37.94  92.56     250.0   
1008    EWR  2013      2   12     2  39.92  28.04  62.21     270.0   
1009    EWR  2013      2   12     3  39.02  26.96  61.63     260.0   
1010    EWR  2013      2   12     4  39.02  26.96  64.29     280.0   

      wind_speed  wind_gust  precip  pressure  visib             time_hour  
1006     6.90468        NaN     0.0    1006.9   10.0  2013-02-12T05:00:00Z  
1007     9.20624        NaN     0.0    1007.2   10.0  2013-02-12T06:00:00Z  
1008    20.71404   25.31716     0.0    1007.8   10.0  2013-02-12T07:00:00Z  
1009  1048.36058        NaN     0.0    1008.3   10.0  2013-02-12T08:00:00Z  
1010    12.65858        NaN     0.0    1008.8   10.0  2013-02-12T09:00:00Z  
Airports:  ['EWR' 'JFK' 'LGA']


In [8]:
# check for duplicate entries for each time and place
dup_counts = w.duplicated(subset=["year","month","day","hour","origin"], keep=False)
duplicates = w[dup_counts].sort_values(["origin","year","month","day","hour"])
duplicates.head(20)


Empty DataFrame
Columns: [origin, year, month, day, hour, temp, dewp, humid, wind_dir, wind_speed, wind_gust, precip, pressure, visib, time_hour]
Index: []

No duplicates were found.

In [9]:
# creating a summary per airport containing maximum observed wind_speed, maximum wind_gust if present, mean wind_speed and count of observations
summary = w.groupby("origin").agg(
    max_wind_speed = ("wind_speed", "max"),
    mean_wind_speed = ("wind_speed", "mean"),
    median_wind_speed = ("wind_speed", "median"),
    max_wind_gust = ("wind_gust", "max"),
    obs_count = ("wind_speed", "count")
).reset_index()
print (summary)

  origin  max_wind_speed  mean_wind_speed  median_wind_speed  max_wind_gust  \
0    EWR      1048.36058         56.38822           13.80936       31.07106   
1    JFK        20.71404         14.38475           14.96014       27.61872   
2    LGA        23.01560         14.96014           14.96014       31.07106   

   obs_count  
0         24  
1         24  
2         24  


The summary shows wind speed observations for EWR, JFK, and LGA airports. While JFK and LGA have similar maximum, mean, and median values, EWR reports an extreme maximum wind speed of 1048.36, which is far above the mean and median values for the same airport. This indicates the presence of an outlier in the dataset that may have resulted from a recording or data entry error.

In [10]:
# remove outlier wind speeds greater than 100
w_clean = w[w["wind_speed"] < 100].copy()

# recompute summary without the outlier
summary_clean = w_clean.groupby("origin").agg(
    max_wind_speed=("wind_speed", "max"),
    mean_wind_speed=("wind_speed", "mean"),
    median_wind_speed=("wind_speed", "median"),
    max_wind_gust=("wind_gust", "max"),
    obs_count=("wind_speed", "count")
).reset_index()

print(summary_clean)

  origin  max_wind_speed  mean_wind_speed  median_wind_speed  max_wind_gust  \
0    EWR        21.86482        13.258987           13.80936       31.07106   
1    JFK        20.71404        14.384750           14.96014       27.61872   
2    LGA        23.01560        14.960140           14.96014       31.07106   

   obs_count  
0         23  
1         24  
2         24  


After removing the wind speed outlier, the summary now reflects consistent maximum, mean, and median wind speeds for all three New York area airports. This provides a more accurate view of the windiest conditions on February 12, 2013.

The answer to question 3 is LaGuardia Airport as it had the highest maximum wind speed on February 13, 2013.